zjit/src/cruby_methods.rs

```diff
@@ -274,6 +275,17 @@ fn inline_string_getbyte(fun: &mut hir::Function, block: hir::BlockId, recv: hir
     None
 }
 
+fn inline_string_append(fun: &mut hir::Function, block: hir::BlockId, recv: hir::InsnId, args: &[hir::InsnId], state: hir::InsnId) -> Option<hir::InsnId> {
+    let &[other] = args else { return None; };
+    if fun.likely_a(other, types::StringExact, state) {
```

Comment:
Let's match the original type check used by `vm_opt_ltlt`. `rb_str_buf_append` is not leaf (it can raise `Encoding::CompatibilityError`), so this should be inlined _if and only if_ the interpreter does that.

`vm_opt_ltlt` checks `RB_TYPE_P(obj, T_STRING)`, so this can (and should) be relaxed to `types::String`. On the other hand, you need to be stricter about `recv`; `recv` should be `types::StringExact`, not a subclass of it.